### ISSUE TITLE
Commanded.aggregate_state does not work when aggregate identity has a prefix

### ISSUE THREAD

**Issue by dvic** [commanded/commanded#516]
I noticed that `Commanded.aggregate_state` does not work with aggregates that have an identity prefix defined in the router. The returned value is an empty aggregate, because it thinks it's a non-started aggregate.

There are few things here I'd like to discuss:

- can this "bug" even be fixed? with only the application and aggregate we don't necessarily have access to the routing info right? so is the "fix" then to change the argument from `aggregate_uuid :: Aggregate.uuid()` to `stream_uuid :: String.t()`? (so that the caller has to prefix the aggregate themselves)

- shouldn't this function raise or return an error when an invalid ID is given instead of an empty aggregate?

**Comment by danhawkins** [comment]
You can just call aggregate_state with the prefixed id, so instead of calling `aggregate_state(AggregateModule, "<uuid>")` use `aggregate_state(AggregateModule, "<prefix>-<uuid>")`

**Comment by dvic** [comment]
But now i'm also a bit confused about whether or not having a router on a commanded application is required or not, it seems that it's technically not required for dispatching but if you don't do this some features don't work  (fetching of state but also `EventAssertions.wait_for_event`)

**Comment by dvic** [comment]
Ok, I have reached the conclusion that basically Commanded expects that the routing info is present on the application because otherwise process managers can't dispatch commands. It does make me wonder though why the public dispatch api even exists on routers directly (this seems to suggest applications don't have to have the routing info per se).

In other words: I think we can assume that the routing information is present on the application and thus I can submit a PR that looks up the prefix from the application composite router (from a compile-time exposed `__registered_prefixes__` function or something like that).

**Comment by slashdotdash** [comment]
>  It does make me wonder though why the public dispatch api even exists on routers directly (this seems to suggest applications don't have to have the routing info per se).

The command router module existed before Commanded added the application module to support multiple applications and event stores. 

It is still possible to dispatch commands directly via a router but you must pass the Commanded application as an option:

```elixir
:ok = BankRouter.dispatch(command, application: BankApp)
```

**Comment by dvic** [comment]
Yeah, that's true, but still, you can't use Process managers. Should I open a PR to document this? Or should we allow process managers to be mounted on routers? I guess they only need the dispatch info?

**Comment by slashdotdash** [comment]
To fix the issue we'd need to add an optional argument to the function to specify the aggregate prefix as it is not possible to get the prefix except via a command registered with a router.

**Comment by dvic** [comment]
> 

If we can't make the assumption generally that the router is embedded on the application, then indeed it's not possible. But if we have the router then we can expose the identities at compile time just like `__registered_commands__` right? Then we could make it work. The only questions (also a more general one): do we want to keep the current behaviour and document it, with this I mean that process managers and `Commanded.aggregate_state` do not work if you don't embed a router on your application. The latter could be fixed by extending the parameter to `stream_uuid` instead of `aggregate_uuid` (i.e., make the caller responsible for prepending the prefix).

**Comment by dvic** [comment]
Regarding the process manager, I see #517 was opened.

**Comment by thetamind** [comment]
Follow me along the journey I took to use `Commanded.aggregate_state` with `identity_prefix`...

### Terminology 

If I understand correctly...
The term `aggregate_uuid` refers to the aggregate identity including the optional prefix.
The [middleware documentation](https://github.com/commanded/commanded/blob/7be1943acdbd313707743bcc9142a888ef869dae/lib/commanded/middleware/pipeline.ex#L30-L36).

> identity_prefix` - an optional prefix to the aggregate's identity.

And the result of composing the optional prefix is also called the "aggregate's identity" or in code labeled `aggregate_uuid`.

The router `identify` macro [configures] how the `aggregate_uuid` is derived from a command.

```elixir
identify BankAccount,
  by: :account_number,
  prefix: "bank-account-"
``` 

The command is dispatched and the default middleware `ExtractAggregateIdentity` reads the config (`identity` and `identity_prefix`) from the `Pipeline`. The value is labeled `aggregate_uuid` before and after the prefix has been applied. Then it is assigned to the middleware pipeline under the key `aggregate_uuid`. From this point forward `prefix` is no longer a concept and `aggregate_uuid` is the full identity for the aggregate and associated stream.

```elixir
def before_dispatch(%Pipeline{} = pipeline) do
    with aggregate_uuid when aggregate_uuid not in [nil, ""] <- extract_aggregate_uuid(pipeline),
         aggregate_uuid when is_binary(aggregate_uuid) <- identity_to_string(aggregate_uuid),
         aggregate_uuid when is_binary(aggregate_uuid) <- prefix(aggregate_uuid, pipeline) do
      assign(pipeline, :aggregate_uuid, aggregate_uuid)
```

The `aggregate_uuid` is used to open the aggregate.

```elixir
{:ok, ^aggregate_uuid} =
      Commanded.Aggregates.Supervisor.open_aggregate(
        application,
        aggregate_module,
        aggregate_uuid
      )
```

And `aggregate_uuid` directly populates `stream_uuid` at the [event store boundary](https://github.com/commanded/commanded/blob/7be1943acdbd313707743bcc9142a888ef869dae/lib/commanded/aggregates/aggregate_state_builder.ex#L41-L61). 

```elixir
case EventStore.stream_forward(
        application,
        aggregate_uuid,
        aggregate_version + 1,
        @read_event_batch_size
      ) do
```

[configures]: https://github.com/commanded/commanded/blob/7be1943acdbd313707743bcc9142a888ef869dae/lib/commanded/commands/router.ex#L87-L90

## Documentation may be unclear

[This paragraph](https://github.com/commanded/commanded/blob/v1.4.1/lib/commanded/commands/router.ex#L94-L102) could produce misunderstanding.

> The **prefix is used as the stream identity** when appending and reading the
>  aggregate's events: `"<identity_prefix><aggregate_uuid>"`. It can be a string or
> a zero arity function returning a string.

The previous paragraph describes the aggregate identity. In the simple case, the identity *is* typically an aggregate id field, hence `aggregate_uuid` here could be an confusing term. Would this be clearer?

`aggregate_uuid = "<identity_prefix><identity>"` or `aggregate identity = "[<prefix>]<identity>"`

And change "The prefix is used as the stream identity" to correct and break into two concepts:

> The aggregate identity is composed of the optional prefix and the identity. 

> The aggregate identity is used as the stream identity when appending and reading the
  aggregate's events
> OR
> The aggregate identity is used to identify the aggregate and is the name of the event stream.
> OR
> The aggregate identity both the `aggregate_uuid` of the aggregate and the `stream_uuid` of the event stream.

So then is the typespec correct because the `aggregate_uuid` and `stream_uuid` are equivalent?


### Access aggregate state during command dispatch is direct
 
1. After command dispatch 

    option `returning: :aggregate_state` or `returning: :execution_result` to grab `aggregate_uuid`. 

2. Dispatch middleware

	Default middleware [ExtractAggregateIdentity](https://github.com/commanded/commanded/blob/v1.4.1/lib/commanded/middleware/extract_aggregate_identity.ex) assigns `:aggregate_uuid` to the middleware `Pipeline` in `before_dispatch`. (`identity` and `identity_prefix` config are there too)

### Access from elsewhere such as a Process manager is not direct

In the simple case the caller joins the prefix string to provide the `aggregate_uuid`. 
In the complex case the identity config supports zero arity functions. The logic to use the config to extract the `aggregate_uuid` is trapped inside `ExtractAggregateIdentity` and can't be reused. 

### My solution

1. Simple case: helper function

```elixir
defmodule Aggregates.UserThing do
  def aggregate_identity(%{user_uuid: user_uuid}) do
    "user-thing-#{user_uuid}"
  end
end

# Can even configure identity in the router now without using prefix key
identify(UserThing, by: &UserThing.aggregate_identity/1)
```

2. Complex case: move identity config outside of the router, clone extract logic

```elixir
defmodule Aggregates.UserThing do
  def identity_config do
    [
      by: :user_uuid,
      prefix: "user-thing-"
    ]
  end
end

# And then use it in the router
identify(UserThing, UserThing.identity_config())

# Copy logic from Commanded.Middleware.ExtractAggregateIdentity into own module
defmodule MyApp.CommandedIdentity do
  @moduledoc """
  Extracts the target aggregate's identity.
  Logic from `Commanded.Middleware.ExtractAggregateIdentity`.
  """

  defmodule Config do
    @enforce_keys [:command, :identity, :identity_prefix]
    defstruct [:command, :identity, :identity_prefix]
  end

  alias MyApp.CommandedIdentity.Config

  def aggregate_uuid(message, opts) when is_list(opts) do
    opts =
      Keyword.new(opts, fn
        {:by, value} -> {:identity, value}
        {:prefix, value} -> {:identity_prefix, value}
      end)
      |> Keyword.put(:command, message)

    config = struct!(Config, opts)
    aggregate_uuid(config)
  end

  def aggregate_uuid(%Config{} = config) do
[...]
  end
end
```

To use it. 

```elixir
aggregate_uuid = MyApp.CommandedIdentity.aggregate_uuid(command_or_query, UserThing.identity_config())
state = CommandedApp.aggregate_state(UserThing, aggregate_uuid)
```

I determined I didn't need the complex case complications. I went with option 1 to query a simple aggregate state.
